manager/src/main/java/io/gingersnapproject/configuration/ConfigurationListener.java

```diff
@@ -0,0 +1,50 @@
+package io.gingersnapproject.configuration;
+
+import java.util.Map;
+
+import javax.enterprise.event.Observes;
+
+import org.eclipse.microprofile.config.ConfigProvider;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.gingersnapproject.k8s.RuleEvents;
+import io.smallrye.config.SmallRyeConfig;
+
+public class ConfigurationListener {
+   SmallRyeConfig smConfig = ConfigProvider.getConfig().unwrap(SmallRyeConfig.class);
+   Configuration config = smConfig.getConfigMapping(Configuration.class);
+
+   static final Logger log = LoggerFactory.getLogger(ConfigurationListener.class);
+   
+   public void onAddRule(@Observes  RuleEvents.LazyRuleAdded ev) {
+      addRule(config.lazyRules(), ev.name(), ev.rule());
+   }
+
+   public void onRemoveRule(@Observes RuleEvents.LazyRuleRemoved ev) {
+      removeRule(config.lazyRules(), ev.name());
+   }
+
+   public void onAddRule(@Observes  RuleEvents.EagerRuleAdded ev) {
+      addRule(config.eagerRules(), ev.name(), ev.rule());
+   }
+
+   public void onRemoveRule(@Observes RuleEvents.EagerRuleRemoved ev) {
+      removeRule(config.eagerRules(), ev.name());
+   }
+
+   public <T> void removeRule(Map<String, T> map, String name) {
+      var removed = map.remove(name);
+      if ( removed == null) {
+         log.warn("Rule("+name+") not present. Not removed.");
+      }
+   }
+
+   public <T> void addRule(Map<String, T> map, String name, T rule) {
+      var ruleInMap = map.putIfAbsent(name, rule);
```

Comment:
Unfortunately we can't do this if no rules are defined on startup, as a `Collections$EmptyMap` is used, so on adding a rule we get:

```java
2023-02-02 12:19:57,170 ERROR [io.fab.kub.cli.inf.imp.cac.SharedProcessor] (-1241025797-pool-7-thread-1) v1/namespaces/test/configmaps failed invoking io.gingersnapproject.k8s.CacheRuleInformer$ConfigMapEventHandler@281a4f8a event handler: null: java.lang.UnsupportedOperationException
	at java.util.Collections$EmptyMap.putIfAbsent(Collections.java:4743)
	at io.gingersnapproject.configuration.ConfigurationListener.addRule(ConfigurationListener.java:44)
	at io.gingersnapproject.configuration.ConfigurationListener.onAddRule(ConfigurationListener.java:21)
	at io.gingersnapproject.configuration.ConfigurationListener_Observer_onAddRule_5dd0ceb17321bcda2b840520d1be2b68954b3d32.notify(Unknown Source)
	at io.quarkus.arc.impl.EventImpl$Notifier.notifyObservers(EventImpl.java:326)
	at io.quarkus.arc.impl.EventImpl$Notifier.notify(EventImpl.java:308)
	at io.quarkus.arc.impl.EventImpl.fire(EventImpl.java:76)
	at io.gingersnapproject.k8s.NotificationManager.lazyRuleAdded(NotificationManager.java:22)
	at io.gingersnapproject.k8s.NotificationManager_ClientProxy.lazyRuleAdded(Unknown Source)
	at io.gingersnapproject.k8s.CacheRuleInformer$LazyRuleEventDispatcher.addEvent(CacheRuleInformer.java:209)
	at io.gingersnapproject.k8s.CacheRuleInformer$ConfigMapEventHandler.lambda$onAdd$0(CacheRuleInformer.java:134)
	at io.gingersnapproject.k8s.CacheRuleInformer$ConfigMapEventHandler.processConfigMapAndSend(CacheRuleInformer.java:127)
	at io.gingersnapproject.k8s.CacheRuleInformer$ConfigMapEventHandler.onAdd(CacheRuleInformer.java:134)
	at io.gingersnapproject.k8s.CacheRuleInformer$ConfigMapEventHandler.onAdd(CacheRuleInformer.java:114)
	at io.fabric8.kubernetes.client.informers.impl.cache.ProcessorListener$AddNotification.handle(ProcessorListener.java:103)
	at io.fabric8.kubernetes.client.informers.impl.cache.ProcessorListener.add(ProcessorListener.java:50)
	at io.fabric8.kubernetes.client.informers.impl.cache.SharedProcessor.lambda$distribute$0(SharedProcessor.java:87)
	at io.fabric8.kubernetes.client.informers.impl.cache.SharedProcessor.lambda$distribute$1(SharedProcessor.java:110)
	at io.fabric8.kubernetes.client.utils.internal.SerialExecutor.lambda$execute$0(SerialExecutor.java:58)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.lang.Thread.run(Thread.java:833)
	at com.oracle.svm.core.thread.PlatformThreads.threadStartRoutine(PlatformThreads.java:705)
	at com.oracle.svm.core.posix.thread.PosixPlatformThreads.pthreadStartRoutine(PosixPlatformThreads.java:202)

```

Reply:
A possible solution is to re-purpose this class as a `RuleManager` bean that  keeps track of Lazy and Eazy rule configurations. On startup the manager will read all statically defined rules from the configuration, then at runtime it will use the `@Observes` events to add/remove runtime rules.

The various components that depend on rules will have to `@Inject` the `RuleManager` instead of `Configuration` and we can use `@Priority` to ensure that the `RuleManager` always receives updates first.